client/extensions/woocommerce/app/settings/shipping/shipping-units.js

```diff
@@ -0,0 +1,137 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { map } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import QuerySettingsProducts from 'woocommerce/components/query-settings-products';
+import {
+	areSettingsProductsLoaded,
+	getWeightUnitSetting,
+	getDimensionsUnitSetting,
+} from 'woocommerce/state/sites/settings/products/selectors';
+import { changeSettingsProductsSetting } from 'woocommerce/state/sites/settings/products/actions';
+import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import FormLabel from 'components/forms/form-label';
+import FormSelect from 'components/forms/form-select';
+
+const ShippingUnits = ( {
+	siteId,
+	loaded,
+	weight,
+	dimensions,
+	translate,
+	changeSetting,
+	onChange,
+} ) => {
+	const weightLabels = unit => {
+		const labels = {
+			oz: translate( 'Ounces' ),
+			lbs: translate( 'Pounds' ),
+			kg: translate( 'Kilograms' ),
+			g: translate( 'Grams' ),
+		};
+		return labels[ unit ] || '';
+	};
+
+	const renderWeightOption = option => {
+		return (
+			<option key={ option } value={ option }>
+				{ `${ option } - ${ weightLabels( option ) }` }
+			</option>
+		);
+	};
+
+	const dimensionsLabels = unit => {
+		const labels = {
+			m: translate( 'Meters' ),
+			cm: translate( 'Centimetres' ),
+			mm: translate( 'Milimetres' ),
+			in: translate( 'Grams' ),
+			yd: translate( 'Yards' ),
+		};
+		return labels[ unit ] || '';
+	};
+
+	const renderDimensionsOption = option => {
+		return (
+			<option key={ option } value={ option }>
+				{ `${ option } - ${ dimensionsLabels( option ) }` }
+			</option>
+		);
+	};
+
+	const onChangeWeight = e => {
+		const setting = Object.assign( {}, weight, { value: e.target.value } );
+		onChange();
+		changeSetting( siteId, setting );
+	};
+
+	const onChangeDimensions = e => {
+		const setting = Object.assign( {}, dimensions, { value: e.target.value } );
+		onChange();
+		changeSetting( siteId, setting );
+	};
+
+	return (
+		<div className="shipping__units">
+			<QuerySettingsProducts siteId={ siteId } />
+			<div className="shipping__weight-select">
+				<FormLabel>{ translate( 'Weight unit' ) }</FormLabel>
+				<FormSelect onChange={ onChangeWeight } value={ weight.value } disabled={ ! loaded }>
+					{ loaded && map( weight.options, renderWeightOption ) }
```

Comment:
Guessing if `loaded` is truthy then `weight.options` will be defined, but should we add a check for `weight` just to be safe here?

Reply:
nm `map` is good about handling nulls in a scenario like this.